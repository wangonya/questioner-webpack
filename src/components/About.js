import React from "react";
import { Container, Row, Col } from "reactstrap";

export const About = () => {
  return (
    <Container>
      <Row className="pt-3">
        <Col data-test="about-page">About Page</Col>
      </Row>
    </Container>
  );
};
