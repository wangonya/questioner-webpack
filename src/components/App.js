import React from "react";
import TopNav from "./Navbar";
function App() {
  return (
    <div>
      <TopNav />
    </div>
  );
}

export default App;
