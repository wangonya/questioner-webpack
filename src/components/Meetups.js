import React, { Component } from "react";
import { connect } from "react-redux";
import {
  Spinner,
  Container,
  Row,
  Col,
  Card,
  CardTitle,
  CardText
} from "reactstrap";

import { fetchMeetups } from "../actions/meetupActions";

export class Meetups extends Component {
  componentDidMount() {
    this.props.fetchMeetups();
  }

  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/static
  static renderLoading() {
    return (
      <div className="pt-5 m-5 d-flex justify-content-center">
        <Spinner type="grow" color="danger" data-test="spinner-test" />
      </div>
    );
  }

  renderMeetups() {
    return (
      <Container>
        <Row className="pt-5">
          {this.props.meetups.map(meetup => (
            <Col sm="4" key={meetup.id} data-test="meetup-cards-test">
              <Card body>
                <CardTitle>
                  <h5>{meetup.title}</h5>
                </CardTitle>
                <CardText>{meetup.details}</CardText>
                <a href="#" className="stretched-link" />
              </Card>
            </Col>
          ))}
        </Row>
      </Container>
    );
  }

  render() {
    return (
      <div>
        {this.props.loading ? Meetups.renderLoading() : this.renderMeetups()}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  meetups: state.meetups.meetups,
  loading: state.meetups.loading
});

export default connect(
  mapStateToProps,
  { fetchMeetups }
)(Meetups);
