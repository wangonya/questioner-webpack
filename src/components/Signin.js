import React, { Component } from "react";
import {
  Spinner,
  Row,
  Col,
  Button,
  FormGroup,
  Label,
  Toast,
  ToastBody,
  ToastHeader
} from "reactstrap";
import { Form, Control } from "react-redux-form";
import { connect } from "react-redux";
import { Redirect } from "react-router";

import { signIn } from "../actions/signinActions";
import { loadingAction } from "../actions/loadingActions";

export class Signin extends Component {
  handleSubmit(data) {
    this.props.loadingAction();
    this.props.signIn(data);
  }

  renderLoading() {
    if (this.props.isSignedIn) {
      return <Redirect to="/" />;
    } else if (this.props.error) {
      return (
        <Row>
          <Col sm="12" md={{ size: 4, offset: 4 }}>
            <div className="p-5 my-2 rounded">
              <Toast data-test="error-test">
                <ToastHeader>Login Failed</ToastHeader>
                <ToastBody>
                  <p>Please check your credentials and try again</p>
                  <p className="text-danger">{this.props.error}</p>
                  <a href="/signin/">Go Back</a>
                </ToastBody>
              </Toast>
            </div>
          </Col>
        </Row>
      );
    }
    return (
      <div className="pt-5 m-5 d-flex justify-content-center">
        <Spinner type="grow" color="danger" data-test="spinner-test" />
      </div>
    );
  }

  renderForm() {
    return (
      <Row>
        <Col sm="12" md={{ size: 4, offset: 4 }}>
          <Form
            model="signin"
            className="pt-5"
            onSubmit={data => this.handleSubmit(data)}
            data-test="form-test"
          >
            <FormGroup>
              <Label for="email">Email</Label>
              <Control.text
                className="form-control"
                model="signin.email"
                type="email"
                placeholder="email"
                required
              />
            </FormGroup>
            <FormGroup>
              <Label for="password">Password</Label>
              <Control.text
                type="password"
                placeholder="password"
                model="signin.password"
                className="form-control"
                required
              />
            </FormGroup>
            <Button>Sign in</Button>
          </Form>
        </Col>
      </Row>
    );
  }

  render() {
    return (
      <div>{this.props.loading ? this.renderLoading() : this.renderForm()}</div>
    );
  }
}

const mapStateToProps = state => ({
  isSignedIn: state.signin.isSignedIn,
  loading: state.loading.loading,
  error: state.error.error
});

export default connect(
  mapStateToProps,
  { signIn, loadingAction }
)(Signin);
