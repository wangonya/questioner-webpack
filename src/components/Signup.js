import React, { Component } from "react";
import {
  Spinner,
  Row,
  Col,
  Button,
  FormGroup,
  Label,
  Toast,
  ToastBody,
  ToastHeader
} from "reactstrap";
import { Form, Control } from "react-redux-form";
import { connect } from "react-redux";
import { Redirect } from "react-router";

import { signUp } from "../actions/signupActions";
import { loadingAction } from "../actions/loadingActions";

export class Signup extends Component {
  handleSubmit(data) {
    this.props.loadingAction();
    this.props.signUp(data);
  }

  renderLoading() {
    if (this.props.isSignedIn) {
      return <Redirect to="/" />;
    } else if (this.props.error) {
      return (
        <Row>
          <Col sm="12" md={{ size: 4, offset: 4 }}>
            <div className="p-5 my-2 rounded">
              <Toast data-test="error-test">
                <ToastHeader>Signup Failed</ToastHeader>
                <ToastBody>
                  <p>Please check your credentials and try again</p>
                  <p className="text-danger">{this.props.error}</p>
                  <a href="/signup/">Go Back</a>
                </ToastBody>
              </Toast>
            </div>
          </Col>
        </Row>
      );
    }
    return (
      <div className="pt-5 m-5 d-flex justify-content-center">
        <Spinner type="grow" color="danger" data-test="spinner-test" />
      </div>
    );
  }

  renderForm() {
    return (
      <Row>
        <Col sm="12" md={{ size: 4, offset: 4 }}>
          <Form
            model="signin"
            className="pt-5"
            onSubmit={data => this.handleSubmit(data)}
            data-test="form-test"
          >
            <FormGroup>
              <Label for="firstname">First Name</Label>
              <Control.text
                className="form-control"
                model="signin.firstname"
                type="text"
                placeholder="First Name"
                required
              />
            </FormGroup>
            <FormGroup>
              <Label for="lastname">Last Name</Label>
              <Control.text
                type="text"
                placeholder="Last Name"
                model="signin.lastname"
                className="form-control"
                required
              />
            </FormGroup>
            <Label for="phone">Phone Number</Label>
            <FormGroup>
              <Control.text
                type="tel"
                placeholder="34545435"
                model="signin.phonenumber"
                className="form-control"
                required
              />
            </FormGroup>
            <FormGroup>
              <Label for="email">Email</Label>
              <Control.text
                className="form-control"
                model="signin.email"
                type="email"
                placeholder="email"
                required
              />
            </FormGroup>
            <FormGroup>
              <Label for="password">Password</Label>
              <Control.text
                type="password"
                placeholder="password"
                model="signin.password"
                className="form-control"
                required
              />
            </FormGroup>
            <Button>Sign up</Button>
          </Form>
        </Col>
      </Row>
    );
  }

  render() {
    return (
      <div>{this.props.loading ? this.renderLoading() : this.renderForm()}</div>
    );
  }
}

const mapStateToProps = state => ({
  isSignedIn: state.signin.isSignedIn,
  loading: state.loading.loading,
  error: state.error.error
});

export default connect(
  mapStateToProps,
  { signUp, loadingAction }
)(Signup);
