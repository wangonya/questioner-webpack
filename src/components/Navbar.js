import React, { Component } from "react";
import { connect } from "react-redux";
import { BrowserRouter as Router, Route } from "react-router-dom";
import {
  Collapse,
  Navbar,
  NavbarToggler,
  NavbarBrand,
  Nav,
  NavItem,
  NavLink
} from "reactstrap";

import Meetups from "./Meetups";
import Signup from "./Signup";
import Signin from "./Signin";
import { About } from "./About";

export class TopNav extends Component {
  constructor(props) {
    super(props);

    this.toggle = this.toggle.bind(this);
    this.state = {
      isOpen: false
    };
  }
  toggle() {
    this.setState({
      isOpen: !this.state.isOpen
    });
  }

  render() {
    return (
      <Router>
        <Navbar color="light" light expand="md">
          <NavbarBrand href="/">Questioner</NavbarBrand>
          <NavbarToggler onClick={this.toggle} />
          <Collapse isOpen={this.state.isOpen} navbar>
            {this.props.isSignedIn || sessionStorage.getItem("token") ? (
              <Nav className="ml-auto" navbar data-test="signed-in">
                <NavItem>
                  <NavLink href="/about/" data-test="about">
                    About
                  </NavLink>
                </NavItem>
                <NavItem>
                  <NavLink
                    href="/signin/"
                    onClick={() => sessionStorage.clear()}
                    data-test="sign-out"
                  >
                    Sign out
                  </NavLink>
                </NavItem>
              </Nav>
            ) : (
              <Nav className="ml-auto" navbar>
                <NavItem>
                  <NavLink href="/signin/">Sign In</NavLink>
                </NavItem>
                <NavItem>
                  <NavLink href="/signup/">Sign up</NavLink>
                </NavItem>
              </Nav>
            )}
          </Collapse>
        </Navbar>
        <Route path="/" exact component={Meetups} />
        <Route path="/about/" exact component={About} />
        <Route path="/signup/" exact component={Signup} />
        <Route path="/signin/" exact component={Signin} />
      </Router>
    );
  }
}

const mapStateToProps = state => ({
  isSignedIn: state.signin.isSignedIn
});

export default connect(mapStateToProps)(TopNav);
