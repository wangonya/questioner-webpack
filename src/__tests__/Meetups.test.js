import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

import { Meetups } from "../components/Meetups";

describe("<Meetups /> component", () => {
  let props;
  let shallowWrapper;
  beforeEach(() => {
    props = {
      meetups: [
        {
          id: 1,
          title: "Another test meetup"
        }
      ]
    };
    props.fetchMeetups = jest.fn();
    shallowWrapper = shallow(<Meetups {...props} />);
  });

  test("renders without crashing", () => {
    const wrapper = renderer.create(<Meetups {...props} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test("fetches meetups on load", () => {
    const spy = jest.spyOn(shallowWrapper.instance().props, "fetchMeetups");
    expect(spy).toHaveBeenCalled();
  });

  test("meetup cards render correctly", () => {
    expect(shallowWrapper.find("[data-test='meetup-cards-test']")).toHaveLength(
      1
    );
  });
});
