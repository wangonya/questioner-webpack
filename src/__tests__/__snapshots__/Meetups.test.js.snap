// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Meetups /> component renders without crashing 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="pt-5 row"
    >
      <div
        className="col-sm-4"
        data-test="meetup-cards-test"
      >
        <div
          className="card card-body"
        >
          <div
            className="card-title"
          >
            <h5>
              Another test meetup
            </h5>
          </div>
          <p
            className="card-text"
          />
          <a
            className="stretched-link"
            href="#"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
